### PR TITLE
bpo-44263: Py_TPFLAGS_HAVE_GC requires tp_traverse

### PR DESCRIPTION
**Pull request by vstinner**
The PyType_Ready() function now raises an error if a type is defined
with the Py_TPFLAGS_HAVE_GC flag set but has no traverse function
(PyTypeObject.tp_traverse).

https://bugs.python.org/issue44263

### CODE REVIEW

**Comment by vstinner** [comment]
This change breaks C extensions which define types with Py_TPFLAGS_HAVE_GC but without tp_traverse, even if instances of the type are never tracked by the GC. Such extensions worked properly previously.

Maybe the check could be restricted to heap types?

If we don't want to break the backward compatibility, we should check in PyObject_GC_Track() if `type->tp_traverse` is not NULL.

Many instances are created by PyType_GenericAlloc() which calls _PyObject_GC_TRACK(). I tried to modify _PyObject_GC_TRACK() to ensure that it's possible to traverse the instance, as I did in PyObject_GC_Track(), but I got a lot of issues. Many types call _PyObject_GC_TRACK() too early, and it's really difficult to change that (it would require a lot of work).

cc @pablogsal @erlend-aasland @corona10

**Comment by vstinner** [comment]
As I explained at https://bugs.python.org/issue44263#msg394799, this PR cannot be merged yet since _testcapi and _decimal don't respect the GC protocol. I chose to create a separated PR to fix them, so it can be backported (and review should be easier): https://github.com/python/cpython/pull/26464

**Comment by vstinner** [comment]
> Many instances are created by PyType_GenericAlloc() which calls _PyObject_GC_TRACK(). I tried to modify _PyObject_GC_TRACK() to ensure that it's possible to traverse the instance, as I did in PyObject_GC_Track(), but I got a lot of issues. Many types call _PyObject_GC_TRACK() too early, and it's really difficult to change that (it would require a lot of work).

My latest attempt in 2020: https://bugs.python.org/issue40142

File: Doc/whatsnew/3.11.rst [hunk 1]
```diff
@@ -180,4 +185,4 @@ Removed
   parameter of functions :func:`~gettext.translation` and
   :func:`~gettext.install` are also removed, since they are only used for
   the ``l*gettext()`` functions.
-  (Contributed by Dong-hee Na and Serhiy Storchaka in :issue:`44235`.)
\ No newline at end of file
+  (Contributed by Dong-hee Na and Serhiy Storchaka in :issue:`44235`.)
```
Comment (erlend-aasland) [inline Doc/whatsnew/3.11.rst:188]: Why this change? :)

**Review by erlend-aasland — APPROVED** [review]
LGTM. I leave it for Pablo and Dong-hee to decide if this should be restricted to heap types only.

**Comment by pablogsal** [comment]
@vstinner is the issue number correct?

File: Doc/whatsnew/3.11.rst (hunk 1 above)
Comment (vstinner) [reply Doc/whatsnew/3.11.rst:188]: Ask my text editor.

**Comment by vstinner** [comment]
> @vstinner is the issue number correct?

I consider that it's to ok reuse https://bugs.python.org/issue44263 My change implements a runtime check for https://github.com/python/cpython/commit/8b55bc3f93a655bc803bff79725d5fe3f124e2f0 "... it *must* implement the GC protocol itself by at least including a slot".

**Comment by vstinner** [comment]
PR rebased on top on the merged PR #26464 fix.

**Comment by pablogsal** [comment]
> I consider that it's to ok reuse https://bugs.python.org/issue44263

:man_shrugging:  I don't know about that,   https://bugs.python.org/issue44263 is a bout a documentation change and this is a breaking change in the C-API, which probably requires its own issue. But I don't feel strongly so you don't need to change anything. I just want to note that I find it a bit odd.

**Review by corona10 — APPROVED** [review]
lgtm

File: Doc/whatsnew/3.11.rst (hunk 1 above)
Comment (vstinner) [reply Doc/whatsnew/3.11.rst:188]: My text editor (vim) likes to add newline at the end of a file.

**Comment by vstinner** [comment]
For the risk of breaking C extensions with this PR: I documented the change in What's New in Python 3.11. It's the bare minimum change that I can do.

@pablogsal: Apart of the bpo number, are you ok with this change?

A GC type with NULL traverse function is likely to crash, but not always, as we saw with the broken _ssl.SSLError exception ;-) This change makes the code more deterministic, it always fail (but don't crash) ;-)

**Review by shihai1991 — APPROVED** [review]
Nice improvement.

**Comment by shihai1991** [comment]
> For the risk of breaking C extensions with this PR: I documented the change in What's New in Python 3.11. It's the bare minimum change that I can do.

Hm. Ptr's pep-652 have redefined the lifetime of stable ABI, so it's not a big probleam, right?
refs: https://www.python.org/dev/peps/pep-0652/#stable-abi

**Comment by pablogsal** [comment]
> Hm. Ptr's pep-652 have redefined the lifetime of stable ABI, so it's not a big probleam, right?

This doesn't change the ABI, it just makes a bug break early :)

**Comment by shihai1991** [comment]
> This doesn't change the ABI, it just makes a bug break early :)

OK. It's make sense :)

**Comment by vstinner** [comment]
Ok, I merged my PR. If it breaks too many C extensions, we can revert it. I hope that it will detect more real bugs than it breaks C extensions for no good reasons :-D
